src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLambdaParameterShouldBeUnnamedCheck.java

```diff
@@ -0,0 +1,285 @@
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LAMBDA_PARAMETER = "unused.lambda.parameter";
+
+    /**
+     * Invalid parents of the lambda parameter identifier.
+     * These are tokens that can not be parents for a lambda
+     * parameter identifier.
+     */
+    private static final int[] INVALID_LAMBDA_PARAM_IDENT_PARENTS = {
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps track of the lambda parameters in a block.
+     */
+    private final Deque<LambdaParameterDetails> lambdaParameters = new ArrayDeque<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.LAMBDA,
+            TokenTypes.IDENT,
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST rootAST) {
+        lambdaParameters.clear();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (ast.getType() == TokenTypes.LAMBDA) {
+            final DetailAST parameters = ast.findFirstToken(TokenTypes.PARAMETERS);
+            if (parameters != null) {
+                // we have multiple lambda parameters
+                TokenUtil.forEachChild(parameters, TokenTypes.PARAMETER_DEF, parameter -> {
+                    final DetailAST identifierAst = parameter.findFirstToken(TokenTypes.IDENT);
+                    final LambdaParameterDetails lambdaParameter =
+                            new LambdaParameterDetails(ast, identifierAst);
+                    lambdaParameters.push(lambdaParameter);
+                });
+            }
+            else if (ast.getChildCount() != 0) {
+                // we are not switch rule and have a single parameter
+                final LambdaParameterDetails lambdaParameter =
+                            new LambdaParameterDetails(ast, ast.findFirstToken(TokenTypes.IDENT));
+                lambdaParameters.push(lambdaParameter);
+            }
+        }
+        else if (isLambdaParameterIdentifierCandidate(ast) && !isLeftHandOfAssignment(ast)) {
+            // we do not count reassignment as usage
+            lambdaParameters.stream()
+                    .filter(parameter -> parameter.getName().equals(ast.getText()))
+                    .findFirst()
+                    .ifPresent(LambdaParameterDetails::registerAsUsed);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        // when leaving a given lambda, check every parameter associated with it
+        while (lambdaParameters.peek() != null
+                    && ast.equals(lambdaParameters.peek().enclosingLambda)) {
```

Comment:
You may not be able to do my previous comment because we don't push everything (please confirm anyways), but one of the conditions does multiple pushes and you only do a single pop. Isn't that an issue?

Reply:
> The comment is misleading now, we aren't leaving only lambdas here. We are leaving IDENTs too, but the equals method is preventing that check.

Comment is removed. actually, it doesn't add much value.

> Why do we need to do the equals check if we have a lambda check instead? Lambdas are guaranteed to be the same order to leave as they came in.


Can you restate I think I don't understand the question well.  but basically, we have the equals check to check where to stop checking the parameters. suppose we have nested lambda parameters each has 3 parameters when we leave the inner one we want to pop and log the top 3 only so we continue to pop unless `ast` (the token we are leaving) is not equal to `enclosingLambda` of the top of the stack at this point we know that we popped all of `ast` parameters

Reply:
> one of the conditions does multiple pushes and you only do a single pop. Isn't that an issue?

what condition? for each `LAMBDA` we push all of its parameters (one or multiple parameters) and when we leave it we pop all of its parameters also. I don't see condition where we push more that pop

Reply:
> pop all of its parameters 

I missed the while loop. I was thinking it was an if statement.

Reply:
> Comment is removed. actually, it doesn't add much value.

Not done.

Reply:
very weird. Done